File: app-ios/VelaWallet/VelaWallet/Features/Wallet/WalletModels.swift
//
//  WalletModels.swift
//  VelaWallet
//
//  Wallet view models (spec 015, data-model.md). Components consume ONLY
//  these display-ready shapes — no service types, no formatting, no
//  fetching (FR-005 / SC-005). A later "real data" feature replaces the
//  fixture layer that builds them and nothing else. Mirrors the web
//  reference `src/lib/wallet/model.ts`.
//

import SwiftUI

/// The nine mobile home states plus the chain-select sheet (H8).
enum MobileStateId: String, CaseIterable, Identifiable {
    case h1, h1s, h2, h3, h4, h5, h6, h7, h7x, h8
    var id: String { rawValue }

    /// Gallery chip label — mock naming, not translatable copy.
    var label: String {
        switch self {
        case .h1: "H1"
        case .h1s: "H1s"
        case .h2: "H2"
        case .h3: "H3"
        case .h4: "H4"
        case .h5: "H5"
        case .h6: "H6"
        case .h7: "H7"
        case .h7x: "H7x"
        case .h8: "H8"
        }
    }
}

struct WalletHeaderModel {
    let name: String
    let addressDisplay: String
    /// Raw seed — IdenticonAvatar normalizes through vela-core (FR-006).
    let identiconSeed: String
}

enum NetworkPillModel {
    case all(dots: [Color], label: String)
    case single(dot: Color, label: String)
}

enum BalanceStateKind {
    case normal, zeroLive, loading, hidden
}

struct BalanceStatusModel {
    enum Kind { case warning, refreshing }
    let kind: Kind
    let text: String
}

struct BalanceModel {
    let label: String
    let currency: String
    let state: BalanceStateKind
    /// e.g. "$1,383" — absent while loading; mask string when hidden.
    var integer: String?
    /// e.g. "28" — rendered de-emphasised after the separator.
    var decimals: String?
    /// zero-live only (实时 · 监听收款中).
    var liveText: String?
    var status: BalanceStatusModel?
    let a11yHide: String
    let a11yShow: String
}

enum ActivityKind {
    case sent, received, dapp
}

struct ActivityRowModel: Identifiable {
    let id = UUID()
    let kind: ActivityKind
    let title: String
    let subtitle: String
    let amount: String
    let unit: String
    let positive: Bool
    let masked: Bool
    let badgeColor: Color
}

struct ActivityGroupModel: Identifiable {
    let id = UUID()
    let label: String
    let rows: [ActivityRowModel]
}

enum AssetFiatModel {
    case value(String)
    /// Orange 无价格 marker (H4).
    case noPrice(String)
    case masked
}

struct AssetRowModel: Identifiable {
    let id = UUID()
    let ticker: String
    let chain: String
    let badgeColor: Color
    let balance: String
    let fiat: AssetFiatModel
    let masked: Bool
}

enum SectionMode {
    case rows, empty, loading
}

struct SectionEmptyModel {
    let title: String
    let caption: String
}

struct SectionModel {
    let title: String
    let action: String
    let mode: SectionMode
    var empty: SectionEmptyModel?
}

enum ChainDot {
    /// 所有网络 row — neutral dot resolved from the theme at render time.
    case all
    case color(Color)
}

struct ChainRowModel: Identifiable {
    let id = UUID()
    let name: String
    let dot: ChainDot
    let count: Int
    let selected: Bool
}

struct ChainSheetModel {
    let title: String
    let rows: [ChainRowModel]
}

struct TabsModel {
    let wallet: String
    let contacts: String
    let explore: String
    let settings: String
}

struct ActionsModel {
    let receive: String
    let send: String
    let scan: String
}

struct WalletHomeModel {
    let state: MobileStateId
    let header: WalletHeaderModel
    let pill: NetworkPillModel
    let balance: BalanceModel
    let actions: ActionsModel
    let activitySection: SectionModel
    let activityGroups: [ActivityGroupModel]
    let assetsSection: SectionModel
    let assetRows: [AssetRowModel]
    let tabs: TabsModel
    var sheet: ChainSheetModel?
    /// 1 or 1.35 — multiplies wallet type roles via walletTextScale (FR-011).
    let textScale: CGFloat
}

File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/wallet/WalletModels.kt
package app.getvela.wallet.feature.wallet

import androidx.compose.runtime.Immutable
import androidx.compose.ui.graphics.Color

/**
 * Wallet view models (spec 015, data-model.md — Android port of the web's
 * `src/lib/wallet/model.ts`).
 *
 * Components consume ONLY these display-ready shapes — no service types, no
 * formatting, no fetching (spec FR-005 / SC-005). A later "real data" feature
 * replaces the fixture layer that builds them and nothing else.
 */

enum class WalletScreenState { H1, H1S, H2, H3, H4, H5, H6, H7, H7X, H8 }

@Immutable
data class WalletHeaderModel(
    val name: String,
    val addressDisplay: String,
    /** Identicon seed (the full address); rendered via core/identicon. */
    val identiconSeed: String,
)

@Immutable
sealed interface NetworkPillModel {
    data class All(val dots: List<Color>, val label: String) : NetworkPillModel
    data class Single(val dot: Color, val label: String) : NetworkPillModel
}

enum class BalanceStateKind { Normal, ZeroLive, Loading, Hidden }

enum class BalanceStatusKind { Warning, Refreshing }

@Immutable
data class BalanceStatusModel(val kind: BalanceStatusKind, val text: String)

@Immutable
data class BalanceModel(
    val label: String,
    val currency: String,
    val state: BalanceStateKind,
    /** e.g. "$1,383" — null while loading; mask dots while hidden. */
    val integer: String? = null,
    /** e.g. "28" — rendered de-emphasised after the separator. */
    val decimals: String? = null,
    val liveText: String? = null,
    val status: BalanceStatusModel? = null,
    val a11yHide: String,
    val a11yShow: String,
)

enum class ActivityKind { Sent, Received, Dapp }

@Immutable
data class ActivityRowModel(
    val kind: ActivityKind,
    val title: String,
    val subtitle: String,
    val amount: String,
    val unit: String,
    val positive: Boolean,
    val masked: Boolean,
    val badgeColor: Color,
)

@Immutable
data class ActivityGroupModel(val label: String, val rows: List<ActivityRowModel>)

@Immutable
sealed interface AssetFiatModel {
    data class Value(val text: String) : AssetFiatModel
    data class NoPrice(val text: String) : AssetFiatModel
    data object Masked : AssetFiatModel
}

@Immutable
data class AssetRowModel(
    val ticker: String,
    val chain: String,
    val badgeColor: Color,
    val balance: String,
    val fiat: AssetFiatModel,
    val masked: Boolean,
)

enum class SectionMode { Rows, Empty, Loading }

@Immutable
data class EmptyStateModel(val title: String, val caption: String)

@Immutable
data class SectionModel(
    val title: String,
    val action: String,
    val mode: SectionMode,
    val empty: EmptyStateModel? = null,
)

@Immutable
data class ChainRowModel(
    val name: String,
    /** null = the all-networks row (neutral dot). */
    val dot: Color?,
    val count: Int,
    val selected: Boolean,
)

@Immutable
data class SheetModel(val title: String, val rows: List<ChainRowModel>)

@Immutable
data class TabsModel(
    val wallet: String,
    val contacts: String,
    val explore: String,
    val settings: String,
)

@Immutable
data class ActionsModel(val receive: String, val send: String, val scan: String)

@Immutable
data class WalletHomeModel(
    val state: WalletScreenState,
    val header: WalletHeaderModel,
    val pill: NetworkPillModel,
    val balance: BalanceModel,
    val actions: ActionsModel,
    val activitySection: SectionModel,
    val activityGroups: List<ActivityGroupModel>,
    val assetsSection: SectionModel,
    val assetRows: List<AssetRowModel>,
    val tabs: TabsModel,
    val sheet: SheetModel? = null,
    /** 1 or 1.35 — multiplies the font scale via LocalDensity (spec FR-011). */
    val textScale: Float = 1f,
)
